bson.go

```diff
@@ -1,10 +1,88 @@
+// Package bson implements encoding and decoding of BSON as defined by https://bsonspec.org/spec.html
+// and https://www.mongodb.com/docs/manual/reference/bson-types/.
+//
+// # Types
+//
+// The following BSON types are supported:
+//
+//	BSON                Go
+//
+//	Double              float64
+//	String              string
+//	Object              *bson.Object
+//	Array               *bson.Array
+//	Binary data         bson.Binary
+//	ObjectId            bson.ObjectID
+//	Boolean             bool
+//	Date                time.Time
+//	Null                bson.NullType
+//	Regular Expression  bson.Regex
+//	32-bit integer      int32
+//	Timestamp           bson.Timestamp
+//	64-bit integer      int64
+//
+// Composite types (Object and Array) are passed by pointers. Scalar types are passed by values.
 package bson
 
 import (
 	"bytes"
+	"fmt"
 	"sort"
 )
 
+type rawObject []byte
+
+type rawArray []byte
+
+type field struct {
+	name  string
+	value any // BSON type, rawObject, or rawArray
+}
+
+type Object struct {
+	fields []field
+}
+
+type Array struct {
+	elements []any // BSON type, rawObject, or rawArray
+}
+
+func (o *Object) MarshalBinary() ([]byte, error) {
+	res := make([]byte, 0, 5)
+
+	for _, f := range o.fields {
+		var t byte
+		var b []byte
+		var err error
+
+		switch value := f.value.(type) {
+		case rawObject:
+			t = 0x03
+			b = value
+		case rawArray:
+			t = 0x04
+			b = value
```

Comment:
That's the main idea – if the value is already in its serialized form, there is no need to re-encode it